File: Feliz/ReactInterop.js
import {
    useEffect as useEffectReact,
    useLayoutEffect as useLayoutEffectReact,
    useDebugValue as useDebugValueReact } from 'react'

export const useEffect = getDisposable => {
    useEffectReact(() => {
        const disposable = getDisposable()
        return () => {
            disposable.Dispose();
        }
    })
}

export const useEffectWithDeps = (getDisposable, deps) => {
    useEffectReact(() => {
        const disposable = getDisposable()
        return () => {
            disposable.Dispose();
        }
    }, deps)
}

export const useLayoutEffect = getDisposable => {
    useLayoutEffectReact(() => {
        const disposable = getDisposable()
        return () => {
            disposable.Dispose();
        }
    })
}

export const useLayoutEffectWithDeps = (getDisposable, deps) => {
    useLayoutEffectReact(() => {
        const disposable = getDisposable()
        return () => {
            disposable.Dispose();
        }
    }, deps)
}

export const useDebugValue = (value, formatter) => useDebugValueReact(value, formatter)

File: Feliz/ReactInterop.fs
namespace Feliz

open Fable.Core

module ReactInterop =
    [<Import("useDebugValue", "./ReactInterop.js")>]
    let useDebugValueWithFormatter<'t>(value: 't, formatter: 't -> string) : unit = jsNative

    [<Import("useEffect", "./ReactInterop.js")>]
    let useEffect (effect: obj) : unit =  jsNative

    [<Import("useEffectWithDeps", "./ReactInterop.js")>]
    let useEffectWithDeps (effect:  obj) (deps: obj) : unit = jsNative

    [<Import("useLayoutEffect", "./ReactInterop.js")>]
    let useLayoutEffect (effect: obj) : unit = jsNative

    [<Import("useLayoutEffectWithDeps", "./ReactInterop.js")>]
    let useLayoutEffectWithDeps (effect:  obj) (deps: obj) : unit = jsNative

